### PR TITLE
DLS 10676

### PR DESCRIPTION
**Pull request by BetaDraconis**
- **DLS-10676: Add wiremock and update bootstrap. Add extra test scenario for FinancialDataConnectorImpl to cover off 404 responses**
- **DLS-10676: Remove println**
- **DLS-10676: Add integration testing framework and auth stubbing. Add integration test scenarios to verify auth stubbing works as expected.**
- **DLS-10676: WIP refactor list returns flow to remove redundant calls and add ITs to verify behaviour**

### CODE REVIEW

File: app/uk/gov/hmrc/cgtpropertydisposals/controllers/actions/AuthenticatedAction.scala [hunk 1]
```diff
@@ -59,7 +59,7 @@ class AuthenticateActionBuilder @Inject() (
         case _                 =>
           Future.successful(forbidden)
       }(carrier, executionContext)
-      .recover { case err: NoActiveSession =>
+      .recover { case _: NoActiveSession =>
```
Comment (LAJW) [inline app/uk/gov/hmrc/cgtpropertydisposals/controllers/actions/AuthenticatedAction.scala]: ⛏️ I'd suggest squashing into original.

File: app/uk/gov/hmrc/cgtpropertydisposals/service/returns/ReturnsService.scala [hunk 2]
```diff
@@ -344,72 +343,85 @@ class DefaultReturnsService @Inject() (
...
+                                               submitReturnWrapper.submitReturnRequest
+                                             )
+                                           )
+                                         )
+      } yield returnSummaries
+
+    desReturnList.flatMap { desReturnListSuccess =>
+      if (desReturnListSuccess.returnList.nonEmpty) { listReturnsResult(desReturnListSuccess) }
+      else { EitherT.fromEither(Right(List.empty[ReturnSummary])) }
+    }
```
Comment (LAJW) [inline app/uk/gov/hmrc/cgtpropertydisposals/service/returns/ReturnsService.scala:366]: 💡 Could've been the cheeky pattern match + EitherT.rightT + type inference:
```scala
    desReturnList.flatMap {
      case DesListReturnsResponse(Nil) => EitherT.rightT(List.empty)
      case DesListReturnsResponse(returnList) => listReturnsResult(returnList)
    }
```

File: app/uk/gov/hmrc/cgtpropertydisposals/service/returns/ReturnsService.scala [hunk 3]
```diff
@@ -344,72 +343,85 @@ class DefaultReturnsService @Inject() (
...
-        _.hasError(SingleDesErrorResponse("NOT_FOUND", "The remote endpoint has indicated that no data can be found."))
+        err => {
+          val hasErr = err.hasError(expectedError)
+          if (hasErr) logger.info(logString)
+          hasErr
+        }
       )
       .merge
 
     response.status === NOT_FOUND && hasNoReturnBody
```
Comment (LAJW) [inline app/uk/gov/hmrc/cgtpropertydisposals/service/returns/ReturnsService.scala:382]: 💡 
```scala
    val isNoDataResponse =
      response.status === NOT_FOUND && response.parseJSON[DesErrorResponse]().exists(_.hasError(expectedError))
    if (isNoDataResponse) logger.info(logString)
    isNoDataResponse
```

File: it/support/WireMockMethods.scala [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+
+      val uriMappingWithHeaders = headers.foldLeft(uriMappingWithQueryParams) { case (m, (key, value)) =>
+        m.withHeader(key, equalTo(value))
+      }
+
+      body match {
+        case Some(extractedBody) => uriMappingWithHeaders.withRequestBody(equalTo(extractedBody))
+        case None                => uriMappingWithHeaders
+      }
+    }
```
Comment (LAJW) [inline it/support/WireMockMethods.scala:58]: 💡 
```scala

    private val mapping =
      method
        .wireMockMapping(urlPathMatching(uri))
        .withQueryParams(queryParams.view.mapValues(matching).toMap.asJava)
        .pipe(headers.foldLeft(_) { case (m, (key, value)) => m.withHeader(key, equalTo(value)) })
        .pipe { mapping =>
          body match {
            case Some(extractedBody) => mapping.withRequestBody(equalTo(extractedBody))
            case None                => mapping
          }
        }

```

File: it/endpoints/GetReturnsControllerISpec.scala [hunk 5]
```diff
@@ -0,0 +1,344 @@
...
+                  "UkAddress" -> Json.obj(
+                    "line1"    -> sampleDesReturn.propertyAddress.addressLine1,
+                    "line2"    -> sampleDesReturn.propertyAddress.addressLine2,
+                    "town"     -> sampleDesReturn.propertyAddress.addressLine3,
+                    "postcode" -> sampleDesReturn.propertyAddress.postalCode
+                  )
+                ),
+                "charges"                -> JsArray.empty,
+                "isRecentlyAmended"      -> true,
+                "expired"                -> false
```
Comment (LAJW) [inline it/endpoints/GetReturnsControllerISpec.scala:329]: ⛏️ Is there no way to construct this with Writes?

File: it/stubs/AuthStub.scala [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+
+  private val enrolment: JsObject = Json.obj(
+    "key"         -> "HMRC-TERS-ORG",
+    "identifiers" -> Json.arr(
+      Json.obj(
+        "key"   -> "SAUTR",
+        "value" -> "7000040245"
+      )
+    )
+  )
```
Comment (LAJW) [inline it/stubs/AuthStub.scala:36]: ⛏️ Don't we have writes to construct this in a less error-prone way?

**Review by LAJW — APPROVED** [review]
Legend:
⛏️ - Nitpick - nice to have, but feel free to ignore.
💡 - Refactoring idea, I couldn't stop myself from trying out, feel free to ignore.
Tests: LGTM
Code history: It looks like you've engaged two different formatters, so there are conflicting formatting changes. If it's not too much work, I'd consider squashing some of the back-and-forth changes.

File: app/uk/gov/hmrc/cgtpropertydisposals/service/returns/ReturnsService.scala [hunk 7]
```diff
@@ -485,46 +497,38 @@ class DefaultReturnsService @Inject() (
...
-    totalCGTLiability: BigDecimal,
-    charges: Option[List[DesCharge]]
-  )
-
-  implicit val chargeFormat: OFormat[DesCharge]                             = Json.format
-  implicit val returnFormat: OFormat[DesReturnSummary]                      = Json.format
+  final case class DesSubmitReturnResponseDetails(chargeReference: Option[String],
+                                                  amount: Option[BigDecimal],
+                                                  dueDate: Option[LocalDate],
+                                                  formBundleNumber: String)
```
[the remark this answers is not in the document]
Comment (BetaDraconis) [reply app/uk/gov/hmrc/cgtpropertydisposals/service/returns/ReturnsService.scala]: Cool, I will remove this

File: app/uk/gov/hmrc/cgtpropertydisposals/controllers/actions/AuthenticatedAction.scala [hunk 8]
```diff
@@ -57,7 +57,6 @@ class AuthenticateActionBuilder @Inject() (
           val user = AuthenticatedUser(credentials.providerId)
           block(new AuthenticatedRequest[A](user, LocalDateTime.now(), carrier, request))
         case _                 =>
-          println("here")
```
[the remark this answers is not in the document]
Comment (BetaDraconis) [reply app/uk/gov/hmrc/cgtpropertydisposals/controllers/actions/AuthenticatedAction.scala]: Whoopsie

File: app/uk/gov/hmrc/cgtpropertydisposals/controllers/actions/AuthenticatedAction.scala (hunk 1 above)
Comment (BetaDraconis) [reply app/uk/gov/hmrc/cgtpropertydisposals/controllers/actions/AuthenticatedAction.scala]: Will do

File: it/endpoints/GetReturnsControllerISpec.scala (hunk 5 above)
Comment (BetaDraconis) [reply it/endpoints/GetReturnsControllerISpec.scala:329]: There is I was just being lazy

File: it/stubs/AuthStub.scala (hunk 6 above)
Comment (BetaDraconis) [reply it/stubs/AuthStub.scala:36]: The credentials model lacks JSON writes to utilise, but the enrolment model does have one so I have replaced this value with an implementation using the appropriate case class, and used the appropriate writes to populate the JSON

**Review by LAJW — APPROVED** [review]

**Pull request merged**